File: 0189-rotate-array/0189-rotate-array.java
class Solution {

    private void reverser(int[] arr, int start, int end) {
      int temp = 0;
      while (start < end) {
        temp = arr[end];
        arr[end--] = arr[start];
        arr[start++] = temp;
      }
    }

    public void rotate(int[] nums, int k) {
      int n = nums.length;
      k %= n; // just to contain within the nums length range

      reverser(nums, 0, n - 1); // complete array
      reverser(nums, 0, k - 1); // from 0 to before k
      reverser(nums, k, n - 1); // from k to last
    }
}

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: list[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def swaps(arr, start, end):
          while start < end:
            arr[start], arr[end] = arr[end], arr[start]
            start += 1
            end -= 1
          return True
        
        k %= len(nums)

        swaps(nums, 0, len(nums) - 1)
        swaps(nums, k, len(nums) - 1)
        swaps(nums, 0, k - 1)
